Add support for Windows IO completions to the portable thread pool

kouvel opened the pull request:
- Added an implementation for BindHandle
- Polling for IO completions is done in batches on separate threads similarly to what is done on Unixes
- Added a high-priority work item queue to have IO completion work items run at higher priority
- Removed the time-sensitive work item queue, used the high-priority queue instead

Test fixes on Mono
- Fixes https://github.com/dotnet/runtime/issues/63185
- Fixes https://github.com/dotnet/runtime/issues/34582

Commit:
Add support for Windows IO completions to the portable thread pool

- Added an implementation for BindHandle
- Polling for IO completions is done in batches on separate threads similarly to what is done on Unixes
- Added a high-priority work item queue to have IO completion work items run at higher priority
- Removed the time-sensitive work item queue, used the high-priority queue instead

kouvel commented:
ASP.NET perf on citrine-win:

| | Before | After | Diff |
|-:|-:|-:|-:|
| PlaintextPlatform | 10,346,490 | 10,627,688 | 2.7% |
| JsonPlatform | 840,775 | 973,058 | 15.7% |
| FortunesPlatform | 387,290 | 428,023 | 10.5% |
| Plaintext | 5,201,727 | 5,197,202 | -0.1% |
| Json | 824,142 | 919,967 | 11.6% |
| Fortunes | 277,752 | 303,691 | 9.3% |

teo-tsirpanis commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/ThreadPool.Windows.cs:
```diff
@@ -405,5 +411,35 @@ internal static unsafe void RequestWorkerThread()
             return registeredWaitHandle;
         }
+
+        private static unsafe void NativeOverlappedCallback(object? obj)
+        {
+            Debug.Assert(obj != null);
+            NativeOverlapped* overlapped = (NativeOverlapped*)(IntPtr)obj;
+            _IOCompletionCallback.PerformSingleIOCompletionCallback(overlapped, 0, 0);
+        }
+
+        [CLSCompliant(false)]
+        [SupportedOSPlatform("windows")]
+        public static unsafe bool UnsafeQueueNativeOverlapped(NativeOverlapped* overlapped)
+        {
+            // OS doesn't signal handle, so do it here (CoreCLR does this assignment in ThreadPoolNative::CorPostQueuedCompletionStatus)
+            overlapped->InternalLow = (IntPtr)0;
+            // Both types of callbacks are executed on the same thread pool
+            return UnsafeQueueUserWorkItem(NativeOverlappedCallback, (IntPtr)overlapped);
+        }
```
You can avoid the `IntPtr`'s boxing allocation:

```suggestion
        private static unsafe void NativeOverlappedCallback(IntPtr ptr)
        {
            NativeOverlapped* overlapped = (NativeOverlapped*)ptr;
            _IOCompletionCallback.PerformSingleIOCompletionCallback(overlapped, 0, 0);
        }

        [CLSCompliant(false)]
        [SupportedOSPlatform("windows")]
        public static unsafe bool UnsafeQueueNativeOverlapped(NativeOverlapped* overlapped)
        {
            // OS doesn't signal handle, so do it here (CoreCLR does this assignment in ThreadPoolNative::CorPostQueuedCompletionStatus)
            overlapped->InternalLow = (IntPtr)0;
            // Both types of callbacks are executed on the same thread pool
            return UnsafeQueueUserWorkItem<IntPtr>(NativeOverlappedCallback, (IntPtr)overlapped, false);
        }
```

kouvel replied:
Fixed in latest commit

teo-tsirpanis commented on src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.IO.Windows.cs:
```diff
@@ -0,0 +1,186 @@
+
+                _nativeEvents =
+                    (Interop.Kernel32.OVERLAPPED_ENTRY*)
+                    Marshal.AllocHGlobal(NativeEventCapacity * Marshal.SizeOf<Interop.Kernel32.OVERLAPPED_ENTRY>());
```
My understanding is that the `NativeMemory` APIs are preferred when allocating memory (correct me if I am wrong), and [`Unsafe.SizeOf` gets optimized](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgBUALWbAEwEsAdgHMA3DXpNWAJQCuAjH3wwWASXkwoEAA4BlDQDc+YGLjG1GzFrPmLlAYQj4tfADYa9UQ8dM0aAbR0MKBkwDAAZbABPCBkMAAoI6NiAaUEeFj0ARxkYG2wXAEoAXRpBDA0BfIY5XGwAMxgGXCCQjAYAeQA1AFEpMIBBAAVB7oARAH1ugDk2KQBNGgBvGgZVpgBmBgBVNQxBoIYXLQcnNwUIAWSYSLM1jYYp7AV9GHaXqBdsLS0YHgAqQ5aN4aT7fX63NbETY7eT7KAMXYVfIQ1ZQ6plBg8ADuUxk+GAGnadQAQpFyrg2FBsAJcA0oLAeCjJEgGBidHwAF6vOoAWTiBQYAF4AHwMHnYKC4Dj5DKc7kAHi6vQGwzGkxm82F/KZzBZbLlRK2/KFoq2NPqynZXKJip6fSGIwm01mcy1BTMAF8aEA=).

```suggestion
                    NativeMemory.Alloc(NativeEventCapacity, Unsafe.SizeOf<Interop.Kernel32.OVERLAPPED_ENTRY>());
```

kouvel replied:
Yea looks like that would be slightly better, fixed in latest commit

Commit:
Translate NtStatus codes to system error codes, fix some other failures, address feedback

marek-safar commented:
@lateralusX please review

lateralusX commented:
@kouvel Should we enable the runtime tests currently disabled on Mono as part of this PR. There are a handful of tests currently labeled with https://github.com/dotnet/runtime/issues/34582 as active issue due to . Then we would get some additional coverage on MonoVM as part of doing this work.

There is also an open issue tied to the same missing functionality, https://github.com/dotnet/runtime/issues/63185, and that should probably be validate by https://github.com/dotnet/runtime/blob/8f87ac7bca2e7b151933f97bc350fec6c46526c2/src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.cs#L17, if we enable that test on MonoVM.

I will build your branch and validate some of these things manual as part of reviewing the MonoVM side of changes.

lateralusX commented:
Manual verification of https://github.com/dotnet/runtime/issues/63185 using this PR on MonoVM on Windows, works.

Commit:
Fix wasm build

Commit:
Try to fix a time-sensitive test, enable some tests on Mono based on #34582

kouvel commented:
@lateralusX, there seem to be a number of tests in various places disabled on Mono for the same reason. I have enabled the tests mentioned in #34582 for now, there are likely more to enable and I can try to do so in a separate commit but I'm not sure I'll find them all.

kouvel commented:
Oh I see that there are a number of other disabled tests referring to #34582, I can start with those

Commit:
Enable some more tests on Mono based on #34582

lateralusX commented:
Running through the rest of the enabled tests locally and all seems to pass.

If we would like to run the Windows library test suite on Mono as part of this PR (for validation), we would need to enable it on PR runs (currently not run on PR's), flip this to get the test suite running on this PR:

https://github.com/dotnet/runtime/blob/a24364a09d9aea98b545f16689a53bafc6b18c14/eng/pipelines/runtime.yml#L1165

Commit:
Miscellaneous fixes

- Renamed config var for number of poller threads to indicate that they are IO pollers
- Removed an unused property
- Reordered IO completion callback arguments for consistency with before and other code paths
- A `FileSystemWatcher` test failed twice, unable to repro locally, tried offloading background work to a separate thread

Commit:
Temporarily enable libraries tests against Mono for this PR

kouvel commented on src/mono/System.Private.CoreLib/src/System/Threading/ThreadPool.Browser.Mono.cs:
```diff
@@ -118,5 +116,36 @@ private static void Callback()
             ThreadPoolWorkQueue.Dispatch();
         }
+
+        private static unsafe void NativeOverlappedCallback(nint overlappedPtr) =>
```
CoreCLR throws PNSE for `UnsafeQueueNativeOverlapped` so I kept that behavior for the portable implementation. It looks like the implementation may have been taken from the NativeAOT implementation, where it had a common implementation between Windows (where that implementation is necessary) and on Unix (where it could throw PNSE). I kept the implementations the same as before for now for those cases that don't use the portable thread pool. So on Mono now it would throw PNSE on Unixes. I think they could all throw PNSE on Unixes, not sure if the minor change in behavior would matter.

danmoseley commented on src/coreclr/System.Private.CoreLib/src/System/Threading/ThreadPool.CoreCLR.Windows.cs:
```diff
@@ -0,0 +1,78 @@
+    {
+        [CLSCompliant(false)]
+        [SupportedOSPlatform("windows")]
+        public static unsafe bool UnsafeQueueNativeOverlapped(NativeOverlapped* overlapped)
+        {
+            if (overlapped == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.overlapped);
+            }
```
```suggestion
        public static unsafe bool UnsafeQueueNativeOverlapped(NativeOverlapped*!! overlapped)
        {
```
at least, I assume that works.

danmoseley replied:
same elsewhere :)

kouvel replied:
Interesting, though I think I'll just leave it as is, unless something is better

stephentoub commented:
> at least, I assume that works.

only once #64720 is merged.

Commit:
Try to fix a time-sensitive test that is occasionally failing

kouvel commented:
Rebased to fix conflicts

Commit:
Note SOS usage of a variable

davidfowl commented:
This makes me so happy 😄

davidfowl commented on src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.IO.Windows.cs:
```diff
@@ -0,0 +1,193 @@
+                {
+                    IsThreadPoolThread = true,
+                    IsBackground = true,
+                    Priority = ThreadPriority.Highest,
```
This is interesting, I assume it was the same in native code?

kouvel replied:
The polling mechanism in this change is similar to how it's done on Unixes and different from what was done before. Before on Windows, there were proc-count threads (at normal priority) polling for IO completions one completion at a time. In this change, there are just a few threads polling for batches of IO completions. Depending on how frequently and how many IO completions occur, a number of polling threads would be scheduled. These few IO poller threads have to compete with all of the active thread pool worker threads to get a time slice to queue work. Hill climbing typically causes more worker threads to be active than the proc count, and many services configure a much higher min worker thread count than the proc count for other reasons. The more worker threads that are active and scheduled to process work, the more time it will take for IO poller threads to get a time slice. Making them higher priority removes those artificial dependencies. Unfortunately, none of the ASP.NET benchmarks reach 100% CPU utilization, so they would not be able to show the difference.

RE @GSPP:
> Is ThreadPriority.Highest really the way to go? This can freeze a Windows system stopping even system services. This can cause a user or administrator to lose the ability to stop the application as well.

I'm aware of that, these threads are few in number and due to inherent limits I don't think they can starve a system. If these threads get to a point of starving a system, they would reach a point where there would not be any more IO completions to handle and would not be scheduled. For priority I think `AboveNormal` is sufficient, I don't think `Highest` is necessary. Since there's a config setting that can control this I've also added a check to ensure that there would not be too many high-priority IO poller threads.

GSPP replied:
This is probably not directly actionable but at least you're aware that something unexpected might be going on with scheduling.

kouvel replied:
> I wanted to see if a high-priority thread becoming ready will immediately deschedule lower priority threads, or if there is a delay.

I was curious about that too. I was hoping that `Highest` priority at least would deschedule a thread sooner but apparently not, so `AboveNormal` is similar. One poller thread should be sufficient but it's not in Kestrel Plaintext because of this delay, and that's unfortunate because more poller threads add to CPU time.

> The IO dequeing threads might wait for an unexpectedly long time and not actually take precedence.

At the very least they would get first dibs on time ahead of other scheduled threads, that's still a bonus. Without that they could be waiting for a very long time in some cases.

dmitry-azaraev replied:
@kouvel it is sounds similar to priority inversion and random priority boosts occurs?

kouvel replied:
> it is sounds similar to priority inversion and random priority boosts occurs (as part of way to fight with inversions, which any OS surely doing)?

I didn't quite follow, can you elaborate?

GSPP replied:
In my testing, I disabled priority boosting for the process. It didn't make a difference.

I found that every 5 seconds, Windows will give a few time slices to starved low-priority threads. I found this behavior described on the internet. This is further evidence that scheduling is not done strictly by priority.

davidfowl commented on src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.IO.Windows.cs:
```diff
@@ -0,0 +1,193 @@
+                1024;
+#endif
+
+            private static readonly Action<Event> ProcessEventDelegate = ProcessEvent;
```
nit: 
```suggestion
            private static readonly Action<Event> ProcessEventDelegate = s_ProcessEvent;
```

davidfowl commented:
@sebastienros Do we have HTTP.sys tests that we can run? We enabled those right?

cc @Tratcher

teo-tsirpanis commented on src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolWorkQueue.cs:
```diff
@@ -851,6 +861,90 @@ public void TransferLocalWork()
+                return;
+            }
+
+            ScheduleForProcessing();
```
I haven't understood this line. We queue the object to the thread pool, but we already running it. Why do it again inside the thread pool?

davidfowl replied:
To parallelize the work. An explanation of this pattern is here https://github.com/dotnet/runtime/blob/9293aab2379d6a122d0cf5ed350356cb23e4d698/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs#L206.

teo-tsirpanis replied:
I see. Thanks for the link.

davidfowl commented on src/libraries/System.Private.CoreLib/src/System/Threading/NativeRuntimeEventSource.PortableThreadPool.NativeSinks.cs:
```diff
@@ -182,6 +195,18 @@ public void ThreadPoolIOEnqueue(RegisteredWaitHandle registeredWaitHandle)
+            {
+                ThreadPoolIODequeue(
+                    (IntPtr)nativeOverlapped,
+                    (IntPtr)OverlappedData.GetOverlappedFromNative(nativeOverlapped).GetHashCode());
```
Is this more expensive than what happened before?

kouvel replied:
Not much, previously the pointer to the managed object was reported, we have switched away from that in other similar cases and used the hash code instead because managed objects can be moved around.

davidfowl commented:
We have one scenario this probably regresses in the HTTP.sys case https://github.com/dotnet/aspnetcore/blob/38bd273d7887dfb408f407d5f6db0d70a8b75e99/src/Servers/HttpSys/src/MessagePump.cs#L213-L216. Where the original goal was to run direction on the IO thread to reduce latency. We should measure though.

GSPP commented:
Is `ThreadPriority.Highest` really the way to go? This can freeze a Windows system stopping even system services. Thread priority works across processes.

sebastienros commented:
You can run benchmarks with httpsys by adding `--variable server=HttpSys` on non-platform benchmarks, e.g., plaintext, json

kouvel commented:
> We have one scenario this probably regresses in the HTTP.sys case https://github.com/dotnet/aspnetcore/blob/38bd273d7887dfb408f407d5f6db0d70a8b75e99/src/Servers/HttpSys/src/MessagePump.cs#L213-L216.

@sebastienros is there a way to set `UnsafePreferInlineScheduling` for the non-platform benchmarks?

kouvel commented:
I collected some numbers with HttpSys and Kestrel with `UnsafePreferInlineScheduling`, it looks like there are a couple of regressions.

### HttpSys with `UnsafePreferInlineScheduling`:

Throughput:

|           | Before  | After   | Diff |
| ---------: | -------: | -------: | ----: |
| Plaintext | 625,165 | 630,711 | 0.9% |
| Json      | 488,564 | 500,372 | 2.4% |
| Fortunes  | 201,763 | 205,465 | 1.8% |

Mean latency:

|           | Before | After | Diff   |
| ---------: | ------: | -----: | ------: |
| Plaintext | 5.35   | 5.13  | \-4.1% |
| Json      | 0.69   | 0.64  | \-7.2% |
| Fortunes  | 13.07  | 12.56 | \-3.9% |

### HttpSys default (without `UnsafePreferInlineScheduling`):

Throughput:

|           | Before  | After   | Diff    |
| ---------: | -------: | -------: | -------: |
| Plaintext | 603,217 | 470,196 | \-22.1% |
| Json      | 483,788 | 491,372 | 1.6%    |
| Fortunes  | 236,315 | 234,131 | \-0.9%  |

Mean latency:

|           | Before | After | Diff   |
| ---------: | ------: | -----: | ------: |
| Plaintext | 4.70   | 4.98  | 6.0%   |
| Json      | 0.63   | 0.64  | 1.6%   |
| Fortunes  | 1.80   | 1.63  | \-9.4% |

Plaintext is a regression in throughput and latency. I looked into this a bit and in the profiles I see that after the change there is a lot more contention in IOCP internals while trying to set up an async read and during `GetQueuedCompletionStatusEx`. Before this change, it looks like the contention is much less, and even looks like some of those async reads are happening synchronously. I'm not sure but I suspect the IOCP threads before may be getting switched out at opportune times, maybe reducing some load on the IOCP system. Reducing the rate at which the high-priority work item queue is parallelized seems to help a bit but throughput is still around 500-510 K. I also tried using a separate IO completion port for each poller but the contention seems to not be specific to a port. I'm not sure at the moment how to resolve this one without using the previous model.

### Kestrel with `UnsafePreferInlineScheduling`:

Throughput:

|           | Before    | After     | Diff |
| ---------: | ---------: | ---------: | ----: |
| Plaintext | 1,703,491 | 1,708,612 | 0.3% |
| Json      | 926,253   | 942,628   | 1.8% |
| Fortunes  | 274,728   | 283,101   | 3.0% |

Mean latency:

|           | Before | After | Diff    |
| ---------: | ------: | -----: | -------: |
| Plaintext | 1.54   | 3.16  | 105.2%  |
| Json      | 6.32   | 0.35  | \-94.5% |
| Fortunes  | 1.66   | 1.26  | \-24.1% |

Plaintext is a regression in latency. On average the latency is a bit higher such as above, but occasionally it is much higher, ~25 or so. The 99th percentile latency is also more frequently much higher than before. I think this may be due to the poller threads occasionally not getting a chance to be scheduled soon enough to keep the queues busy. I'm thinking maybe something like `DOTNET_SYSTEM_NET_SOCKETS_INLINE_COMPLETIONS` could be implemented, as it's a similar issue, or for now the config switch could be used to switch back to the previous IO thread pool if necessary.

Commit:
Address feedback

davidfowl commented:
> I'm thinking maybe something like DOTNET_SYSTEM_NET_SOCKETS_INLINE_COMPLETIONS could be implemented, as it's a similar issue, or for now the config switch could be used to switch back to the previous IO thread pool if necessary.

I would prefer to have this escape hatch flag to in the managed implementation rather than have people flip back to the old thread pool. Though I guess that is a decent temporary work around for this advanced scenario.

We may also have to adjust this

> I also tried using a separate IO completion port for each poller but the contention seems to not be specific to a port. I'm not sure at the moment how to resolve this one without using the previous model.

I was going to suggest this. This is what we did with Libuv, one completion port per poller thread and sockets were assigned to completion ports by doing a round robin.

davidfowl commented:
cc @NGloreous

kouvel commented:
> I would prefer to have this escape hatch flag to in the managed implementation rather than have people flip back to the old thread pool. Though I guess that is a decent temporary work around for this advanced scenario.

For now I'm thinking I'll go ahead and add the same `DOTNET_SYSTEM_NET_SOCKETS_INLINE_COMPLETIONS` config setting and implement it in a somewhat similar, simple fashion as on Unixes (since it's already documented and seems to work in the unsafe cases). If for some reason the old model is necessary, that config would be available too for now and maybe based on usage/need it could be expanded in the future to be more similar to before.

> I was going to suggest this. This is what we did with Libuv, one completion port per poller thread and sockets were assigned to completion ports by doing a round robin.

I didn't see any improvements in using multiple IOCPs, the contention I saw was similar with one or multiple IOCPs, so figured one would suffice. With multiple IOCPs and round-robin binding, more poller threads may have to wake up more frequently.

Commit:
Revert "Temporarily enable libraries tests against Mono for this PR"

This reverts commit 9a73d8e6fbac6d3bb9d70d6aab8b8b0de5b3b6c3.

kouvel commented:
Rebased to fix conflicts

Commit:
Fix build after rebase

Commit:
Add checks for some DOTNET vars from Net.Sockets and implement inlining IO completion callbacks

janvorli reviewed (approved):
LGTM, thank you!

kouvel commented:
Looks like the remaining CI failures are unrelated failures also happening elsewhere.

mconnew commented:
I just learnt about this change/PR. On Windows, in the previous IO thread pool implementation, threads which were bound to an IOCP have a different behavior than normal threads. In the [Threads and Concurrency](https://docs.microsoft.com/en-us/windows/win32/fileio/i-o-completion-ports#threads-and-concurrency) section of the win32 documentation it explains about limiting the number of runnable threads, and .NET would limit it to the number of CPU cores. This has the effect that work executing on an IO thread could not be pre-emptively context switched by another one (within the IO thread pool). Posting the work to a regular thread which has higher priority than a regular thread but without this runnable concurrency limit is going to reduce performance in some scenarios. Unless that thread pool is hard limited in the number of threads in it (and not modifiable with ThreadPool.SetMaxThreads), there's going to be more context switching which introduces more overhead. But if you do that and you execute work on one of the threads which blocks, you don't have any extra threads which can be woken up to complete more work so that's not an option (and you deadlock if the blocking thread is waiting on some other work that's waiting).  
I suspect this change will see some improvements for asp.net core, but that's because it doesn't take advantage of the IOCP capability. I would expect any library which is purposefully utilizing IO threads to parallelize work to see a regression with the cost of context switching. If a library (such as asp.net core) wishes to offload work to a high priority regular thread pool, it should do so itself. I think it's a bad idea to effectively remove IO threads from the runtime and offload the work to a non IOCP thread pool automatically.  
We've been performance testing CoreWCF and comparing Linux with Windows. We have special code to utilize IO threads on Windows an in efficient manner. On Linux, when perf testing with all that code disabled (just let the regular thread pool handle the work) we see measurably worse performance on Linux than Windows. The "known wisdom" is that Linux is faster than Windows for networking, but when correctly utilizing IOCP, Windows can have better throughput than Linux.

kouvel commented:
@mconnew, fair point. It may be possible to limit the IOCP concurrency for worker threads too on Windows since they also use an IOCP, perhaps an option could be exposed through config, it may be interesting to try although there may be some downsides to using it in some cases.

The IOCP concurrency limit applies only so long as threads don't block for any reason. If for instance there is any lock contention that causes one thread to wait (even in a spin-lock with `Sleep(1)`), another thread will be released by the IOCP, and after the lock becomes available there would be one extra thread running. So any sort of frequent lock contention or slightly blocking operation could cause more threads than the concurrency limit to be active despite the concurrency limit, until the extra threads wait for more IO completions.

Extra context switching can also occur between IO threads and worker threads when both are being used, which is not uncommon. Reusing one pool for both types of work can reduce the amount of context switching in those typical cases.

There is another difference between Windows and Linux in how IO completions are processed. Before this change on Windows, IO completions are processed in parallel to worker threads. After this change on Windows, IO completions are queued to the worker thread pool at higher priority. On Linux, IO completions are queued as normal work items to the worker thread pool, and a long work queue could artificially delay IO completions from being processed. There are some other differences in how wait completions are processed.

I'd like to understand what kind of effect this change would have in CoreWCF's performance. I'm also curious which differences are causing it to perform slower on Linux. Can you point me to the performance tests and instructions for running them?

davidfowl commented:
> After this change on Windows, IO completions are queued to the worker thread pool at higher priority. On Linux, IO completions are queued as normal work items to the worker thread pool, and a long work queue could artificially delay IO completions from being processed.

@kouvel feels like we should make this change on linux as well. Another benefit to moving the polling logic into the runtime.

kouvel commented:
> @kouvel feels like we should make this change on linux as well. Another benefit to moving the polling logic into the runtime.

I think it would be good to do at least for quasi-correctness, if not for perf.  Doing so may need some APIs such as https://github.com/dotnet/runtime/issues/61273 or similar (or at least an internal way to get to those APIs).

davidfowl commented:
Should re-open that issue for .NET 8?

kouvel commented:
> Should re-open that issue for .NET 8?

Maybe the first option in https://github.com/dotnet/runtime/issues/61273#issuecomment-963608450 could be used for starters.

davidfowl commented:
Option1b would be to expose `ThreadPool.UnsafeQueueHighPriorityWorkItemInternal` so the sockets assembly and use that instead. That wouldn't require moving the polling loop, but it doesn't get us any closer to making those new public APIs.

mconnew commented:
@kouvel, I could be wrong on this, but my understanding was that when a thread goes into an alertable state (eg the sleep(1) you mentioned), it yields the CPU and the kernel will activate a dormant thread. But that thread which went alertable won't get CPU time again until another thread goes alertable and yields up the CPU? Is this not the case? Does it only limit releasing a thread waiting for a completion from the IOCP if there are more than the configured concurrency limit executing?

The model I would like to see is to have an instantiable ThreadPool class. This class would have the option on Windows to create an IOCP for the instance, as well as options to specify OS thread priority for threads created by the pool. It would have a helper method to bind to a Socket which on Windows would either bind the socket to the ThreadPool instances IOCP or bind to the ambient IO thread pool but have the completion be dispatched to the ThreadPool instance (similar to the new IOCP model) if IOCP hasn't been configured. On Linux it would set up a singleton polling loop in the ThreadPool instance and dispatch ready notifications to the ThreadPool work queue.  

The ThreadPool would have public helper properties which can return a SynchronizationContext and a TaskScheduler to make it easy to run code in the ThreadPool using methods like TaskFactory.StartNew. It would also have its own QueueUserWorkItem method to easily dispatch delegates to it.   

I think this model would satisfy being able to keep the original behavior that WCF/CoreWCF uses to do pseudo cooperative multitasking, make it really easy to have all IO operations initiated and completed in the pool, and it would enable the current new IO thread pool model which improves performance for some scenarios, and it doesn't require compromising.  

I think we're getting to the point where having two implicit built-in thread pools with hard coded behavior is becoming a limiting factor.

davidfowl commented:
@mconnew You should file another issue and write up an API proposal. It sounds like you basically want a way to spin up your own poller for network IO (I wouldn't call this a threadpool).

kouvel commented:
> I could be wrong on this, but my understanding was that when a thread goes into an alertable state (eg the sleep(1) you mentioned), it yields the CPU and the kernel will activate a dormant thread. But that thread which went alertable won't get CPU time again until another thread goes alertable and yields up the CPU? Is this not the case?

It looks like after the sleep expires, that thread is scheduled normally, even if that would exceed the configured concurrency limit. I didn't see a difference in behavior between alertable and non-alertable waits, both seem to allow more than the concurrency limit to run simultaneously. Waiting on `GetQueuedCompletionStatus` does appear to block the extra threads based on the concurrency limit.

> The model I would like to see is to have an instantiable ThreadPool class.

That may be reasonable if the underlying thread resources are shared to a large degree. I would be concerned about having too many thread pools that amount to having too many threads underneath. There may be other advantages to having instantiable thread pools, such as for configuring an instance for how those work items run.

> I think this model would satisfy being able to keep the original behavior that WCF/CoreWCF uses to do pseudo cooperative multitasking, make it really easy to have all IO operations initiated and completed in the pool, and it would enable the current new IO thread pool model which improves performance for some scenarios, and it doesn't require compromising.

I'd like to understand this more. Is there a performance drop with this change, or between .NET 6 and .NET 7? If so, I'd like to understand better what is causing it and see some perf profiles. If it really is a separate pool of IO completion threads that is helping in the scenario, there may be options to consider, and it would help to understand the reason for the difference to determine which options may be fruitful.
